### PR TITLE
Implemented gRPC client for heimdall

### PR DESCRIPTION
**Pull request by 0xKrishna**
This PR contains :

1. gRPC client implementation for the Heimdall server
2. Implemented all the methods in the current Heimdall interface

To use gRPC you have to pass gRPC server address using
```--bor.heimdallgRPC=<address>```
If you give the above flag then it will use gRPC, Otherwise it will continue to use REST APIs.

Default port in Heimdall is `3132`
So for local the gRPC address is `localhost:3132`.

https://polygon.atlassian.net/browse/POS-648

### CODE REVIEW

**Review by JekaMas — APPROVED** [review]

File: consensus/bor/heimdallgrpc/client.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+type HeimdallGRPCClient struct {
+	conn   *grpc.ClientConn
+	client proto.HeimdallClient
+}
+
+func NewHeimdallGRPCClient(address string) *HeimdallGRPCClient {
+	conn, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Error("Failed to connect to Heimdall gRPC server: %v", err)
+		panic(err)
```
Comment (manav2401) [inline consensus/bor/heimdallgrpc/client.go:24]: Can we avoid `panic`?

File: consensus/bor/heimdallgrpc/span.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+func (h *HeimdallGRPCClient) Span(ctx context.Context, spanID uint64) (*span.HeimdallSpan, error) {
+	req := &proto.SpanRequest{
+		ID: spanID,
+	}
+
+	res, err := h.client.Span(context.Background(), req)
+	if err != nil {
+		return nil, err
```
Comment (manav2401) [inline consensus/bor/heimdallgrpc/span.go:20]: Won't we need a retry mechanism, specifically for `span`? Also, I think it's good to have it also for state-sync & checkpoint methods.

File: consensus/bor/heimdallgrpc/state_sync.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/bor/clerk"
+
+	proto "github.com/maticnetwork/polyproto/heimdall"
+)
+
+func (h *HeimdallGRPCClient) StateSyncEvents(ctx context.Context, fromID uint64, to int64) ([]*clerk.EventRecordWithTime, error) {
+	eventRecords := make([]*clerk.EventRecordWithTime, 0)
+
+	req := &proto.StateSyncEventsRequest{
```
Comment (JekaMas) [inline consensus/bor/heimdallgrpc/state_sync.go:15]: I am more than sure that it's possible to remove all this manual unmarshal. By the time, as we discussed, let's keep it as it as. Could you create a task for bor and heimdall to find a bit better solution? We can rethink that place later.

File: consensus/bor/heimdallgrpc/span.go (hunk 2 above)
Comment (JekaMas) [reply consensus/bor/heimdallgrpc/span.go:20]: Why you use blank context here?

File: consensus/bor/heimdallgrpc/state_sync.go (hunk 3 above)
Comment (JekaMas) [reply consensus/bor/heimdallgrpc/state_sync.go:15]: @0xKrishna could we update PRs (Heimdall and Bor), fix CI and merge gRPC PRs?

File: consensus/bor/heimdallgrpc/client.go (hunk 1 above)
Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/client.go:24]: Can use `log.Fataf`
I will check more options.

File: consensus/bor/heimdallgrpc/span.go (hunk 2 above)
Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/span.go:20]: Fixed the context

Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/span.go:20]: Good point regarding retry, I will check

File: consensus/bor/heimdallgrpc/state_sync.go (hunk 3 above)
Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/state_sync.go:15]: Lint and CI and fixed, Working on few comments above

File: consensus/bor/heimdallgrpc/client.go (hunk 1 above)
Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/client.go:24]: Using `log.Fatal` now, Will push soon.

File: consensus/bor/heimdallgrpc/span.go (hunk 2 above)
Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/span.go:20]: Fixed the context, Looking to implement retry mechanism.

Comment (0xKrishna) [reply consensus/bor/heimdallgrpc/span.go:23]: Implemented retry middleware, Now need to add supported errors on heimdall side.

**Review by 0xsharma — APPROVED** [review]

**Review by manav2401 — APPROVED** [review]

**Pull request merged**